java/ql/lib/semmle/code/java/Expr.qll

```diff
@@ -2177,3 +2405,111 @@ class ValueDiscardingExpr extends Expr {
     not this.getType() instanceof VoidType
   }
 }
+
+/** A Kotlin `when` expression. */
+class WhenExpr extends Expr, StmtParent, @whenexpr {
+  override string toString() { result = "when ..." }
+
+  override string getHalsteadID() { result = "WhenExpr" }
+
+  override string getAPrimaryQlClass() { result = "WhenExpr" }
+
+  /** Gets the `i`th branch. */
+  WhenBranch getBranch(int i) { result.isNthChildOf(this, i) }
+
+  /** Holds if this was written as an `if` expression. */
+  predicate isIf() { when_if(this) }
+}
+
+/** A Kotlin `when` branch. */
+class WhenBranch extends Stmt, @whenbranch {
+  /** Gets the condition of this branch. */
+  Expr getCondition() { result.isNthChildOf(this, 0) }
+
+  /** Gets the result of this branch. */
+  Stmt getRhs() { result.isNthChildOf(this, 1) }
+
+  /** Gets a result expression of this `when` branch. */
+  Expr getAResult() { result = getAResult(this.getRhs()) }
+
+  /** Holds if this is an `else` branch. */
+  predicate isElseBranch() { when_branch_else(this) }
+
+  /** Gets the `when` expression this is a branch of. */
+  WhenExpr getWhenExpr() { this = result.getBranch(_) }
+
+  override string toString() { result = "... -> ..." }
+
+  override string getAPrimaryQlClass() { result = "WhenBranch" }
+}
+
+// TODO: This might need more cases. It might be better as a predicate
+// on Stmt, overridden in each subclass.
+private Expr getAResult(Stmt s) {
+  result = s.(ExprStmt).getExpr() or
+  result = getAResult(s.(BlockStmt).getLastStmt())
+}
+
+/** A Kotlin `::class` expression. */
+class ClassExpr extends Expr, @getclassexpr {
+  /** Gets the expression whose class is being returned. */
+  Expr getExpr() { result.isNthChildOf(this, 0) }
+
+  override string toString() { result = "::class" }
+
+  override string getAPrimaryQlClass() { result = "ClassExpr" }
+}
+
+/**
+ * An statement expression.
```

Comment:
**A** statement.

Reply:
Thanks! I've made a PR for these here: https://github.com/github/codeql/pull/9136